Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,399 @@
+<configuration>
+  <system.web>
+    <authorization>
+      <deny users=" + "\"*\"" + @"/>
+    </authorization>
+  </system.web>
+<configuration>";
+
+                        File.WriteAllText(Path.Combine(dotGitPath, "web.config"), content);
+                    }
+
+                    // Server env does not support interactive cred prompt; hence, we intercept any credential provision
+                    // for git fetch/clone with http/https scheme and return random invalid u/p forcing 'fatal: Authentication failed.'
+                    using (tracer.Step("Configure git-credential"))
+                    {
+                        FileSystemHelpers.EnsureDirectory(Path.GetDirectoryName(GitCredentialHookPath));
+
+                        string content = @"#!/bin/sh
+if [ " + "\"$1\" = \"get\"" + @" ]; then
+      echo username=dummyUser
+      echo password=dummyPassword
+fi" + "\n";
+
+                        File.WriteAllText(GitCredentialHookPath, content);
+
+                        repo.Config.Set("credential.helper", string.Format("!'{0}'", GitCredentialHookPath));
+                    }
+                }
+                using (tracer.Step("Setup post receive hook"))
+                {
+                    FileSystemHelpers.EnsureDirectory(Path.GetDirectoryName(PostReceiveHookPath));
+
+                    string content = @"#!/bin/sh
+read i
+echo $i > pushinfo
+" + KnownEnvironment.KUDUCOMMAND + "\n";
+
+                    File.WriteAllText(PostReceiveHookPath, content);
+                }
+
+                // NOTE: don't add any new init steps after creating the post receive hook,
+                // as it's also used to mark that Init was fully executed
+            }
+        }
+
+        private string GitCredentialHookPath
+        {
+            get
+            {
+                return Path.Combine(RepositoryPath, ".git", "hooks", "git-credential-invalid.sh");
+            }
+        }
+
+        public ChangeSet GetChangeSet(string id)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                LibGit2Sharp.Commit commit = repo.Lookup<LibGit2Sharp.Commit>(id);
+
+                if (commit == null)
+                {
+                    return null;
+                }
+
+                return new ChangeSet(commit.Sha, commit.Author.Name, commit.Author.Email, commit.Message, commit.Author.When);
+            }
+        }
+
+        public void AddFile(string path)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Stage(path);
+            }
+        }
+
+        public bool Commit(string message, string authorName, string emailAddress)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                try
+                {
+                    var changes = repo.RetrieveStatus().Select(c => c.FilePath);
```

Comment:
Updated, thanks! Just to make sure I'm actually understanding this, this is basically to reduce the processing that libgit2 has to do to figure out the list of changed files, but `Commit()` will still do the right thing when writing the history, right?